### PR TITLE
Added `Async communication`

### PR DESCRIPTION
**Pull request by boburjon1504**

### CODE REVIEW

File: src/CRM.ServerApp/Crm.Api/Controllers/NotificationsController.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+public class NotificationsController(
+    INotificationOrchestrationService<EmailMessage> emailSender,
+    INotificationOrchestrationService<SmsMessage> smsSender
+    ) : ControllerBase
+{
+    [HttpPost("email")]
+    public async ValueTask<IActionResult> SendEmail([FromBody] EmailMessage emailMessage)
+    {
+        return Ok(await emailSender.SendAsync(emailMessage, cancellationToken: HttpContext.RequestAborted));
+    }
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Api/Controllers/NotificationsController.cs:18]: add line

File: src/CRM.ServerApp/Crm.Application/Notifications/Models/SmsMessage.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Crm.Application.Notifications.Models;
+
+public class SmsMessage : NotificationMessage
+{
+
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Application/Notifications/Models/SmsMessage.cs:5]: no line

File: src/CRM.ServerApp/Crm.Domain/Entities/NotificationTemplate.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using Crm.Domain.Common.Auditables;
+using Crm.Domain.Enums;
+
+namespace Crm.Domain.Entities;
+public abstract class NotificationTemplate : SoftDeletedEntity
+{
+
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Domain/Entities/NotificationTemplate.cs:7]: no line

File: src/CRM.ServerApp/Crm.Domain/Entities/SmsHistory.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Crm.Domain.Entities;
+
+public class SmsHistory : NotificationHistory
+{
+    public SmsHistory()
+    {
+        Type = Enums.NotificationType.Sms;
+    }
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Domain/Entities/SmsHistory.cs:8]: add line

File: src/CRM.ServerApp/Crm.Domain/Entities/SmsHistory.cs [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+    public string SenderPhoneNumber { get; set; }
+    public string ReceiverPhoneNumber { get; set; }
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Domain/Entities/SmsHistory.cs:10]: add line

**Review by MuhammadrizoVositov — APPROVED** [review]

File: src/CRM.ServerApp/Crm.Domain/Enums/NotificationType.cs [hunk 6]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Crm.Domain.Enums;
+public enum NotificationType
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Domain/Enums/NotificationType.cs:2]: add line

File: src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Brokers/AsyncChannelCreatorBroker.cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+using Crm.Infrastructure.Common.Settings;
+using Microsoft.Extensions.Options;
+using RabbitMQ.Client;
+
+namespace Crm.Infrastructure.Common.Notifications.Brokers;
+
+public class AsyncChannelCreatorBroker : IAsyncChannelCreatorBroker
+{
+    private readonly IConnection _connection;
+    public AsyncChannelCreatorBroker(IOptions<RabbitMqSettings> options)
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Brokers/AsyncChannelCreatorBroker.cs:11]: add line

File: src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Brokers/AsyncChannelCreatorBroker.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    {
+        var factory = new ConnectionFactory
+        {
+            HostName = options.Value.Host,
+            //Port = options.Value.Port,
+        };
+        _connection = factory.CreateConnection();
+    }
+    public IChannel CreateChannel()
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Brokers/AsyncChannelCreatorBroker.cs:20]: add line

File: src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailHistoryService.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
+﻿using Crm.Application.Common.Extensions;
+using Crm.Application.Common.Query;
+using Crm.Application.Notifications.Services;
+using Crm.Domain.Entities;
+using Crm.Persistence.Repositories.NotificationsRepositories.Interfaces;
+using FluentValidation;
+using Microsoft.EntityFrameworkCore;
+
+namespace Crm.Infrastructure.Common.Notifications.Services;
+public class EmailHistoryService( INotificationHistoryRepository<EmailHistory> historyRepository,
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailHistoryService.cs:10]: add line

Comment (Ozodbek4) [reply src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailHistoryService.cs:10]: for namespace

File: src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailOrchestrationService.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+                    queuName: "email-message",
+                    exchangeName: ""
+                    );
+            }
+
+            return history.IsSuccessful;
+        };
+        return await sender.GetValueAsync();
+
+
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailOrchestrationService.cs:64]: no line

Comment (Ozodbek4) [reply src/CRM.ServerApp/Crm.Infrastructure/Common/Notifications/Services/EmailOrchestrationService.cs:64]: correct lines

File: src/CRM.ServerApp/Crm.Infrastructure/Common/Settings/RabbitMqSettings.cs [hunk 11]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Crm.Infrastructure.Common.Settings;
+
+public class RabbitMqSettings
+{
+    public string Host { get; set; } = default!;
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.Infrastructure/Common/Settings/RabbitMqSettings.cs:5]: add line

File: src/CRM.ServerApp/Crm.NotificationApi/Auditables/Entity.cs [hunk 12]
```diff
@@ -0,0 +1,5 @@
+﻿namespace Crm.NotificationApi.Auditables;
+public class Entity : IEntity
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.NotificationApi/Auditables/Entity.cs:2]: add line

File: src/CRM.ServerApp/Crm.NotificationApi/Auditables/IEntity.cs [hunk 13]
```diff
@@ -0,0 +1,5 @@
+﻿namespace Crm.NotificationApi.Auditables;
+public interface IEntity
```
Comment (Ozodbek4) [inline src/CRM.ServerApp/Crm.NotificationApi/Auditables/IEntity.cs:2]: add line

**Review by Ozodbek4 — CHANGES_REQUESTED** [review]

**Review by boburjon1504 — COMMENTED** [review]
resolved

**Review by Ozodbek4 — APPROVED** [review]

**Review by MuhammadrizoVositov — APPROVED** [review]

**Pull request merged**